hazelcast/src/main/java/com/hazelcast/spi/impl/proxyservice/impl/operations/PostJoinProxyOperation.java

```diff
@@ -125,7 +132,8 @@ public int getClassId() {
         @Override
         public void run() {
             try {
-                registry.createProxy(proxyInfo.getObjectName(), proxyInfo.getSource(), true, true);
+                registry.createProxy(proxyInfo.getObjectName(), proxyInfo.getSource(), true, true,
```

Comment:
What if the proxies were propagated while the cluster version is 4.0 and then we switched to cluster version 4.1? Where do we react to this cluster version change to update the tenant control?

Reply:
Not sure this can be resolved, or should be, since 4.0 didn't work correctly anyway

Reply:
It might be the case, but here's what will happen on RU from 4.0 to 4.1. A user is running 4.0 cluster with some proxies. They start RU, shut down some 4.0 members, add some 4.1 members, and perhaps add some more proxies in the meantime. They complete RU and switch to cluster version 4.1. At this point, none of the proxies have tenant control since we had to avoid copying it because of serialization concerns.

What I'm thinking is that we should either avoid sending tenant control over the wire and rely on some tenant control on each node locally based on the configuration and the node version, or implement `ClusterVersionListener#onClusterVersionChange` and react on the cluster version change to distribute the tenant control across the cluster once the cluster has been upgraded to 4.1, which seems prone to race conditions as it would be asynchronous.

Reply:
tenant control **absolutely has** to be sent over the wire. This is one of the basic requirements. Again, there is lots of things you don't see that are implemented in Payara and may not be obvious from just the Hazelcast part.

As discussed previously I will try to move it out of proxy and into node engine.

All of this is non-existent problem anyway because nobody is using tenant control in 4.0 because it doesn't work correctly anyway.

Reply:
To answer your question directly, if tenant control is not found (null) it won't be used in which case you will get classloader errors, which would happen in 4.0 anyway since tenant control isn't working correctly.
So, the situation is improved in all cases, and breaks in none.

Reply:
Ok, I'll reassess the code from the standpoint of having to migrate tenant control. I think it might be ok then, and we might further be comfortable with the incompatibility with RU, if we made this whole tenant control SPI internal. But I have to think some more about the consequences of something like that.

Reply:
Either way it's fine with me